File: automation-test-selenium/src/main/java/org/suntower/core/BasePage.java
package org.suntower.core;

import java.net.URI;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;
import java.util.function.BooleanSupplier;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.TimeoutException;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.support.ui.Select;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.suntower.helpers.text.TextNormalizeHelper;
import org.suntower.pages.components.SweetAlertComponent;

public class BasePage {
  protected final WebDriver driver;
  protected final AppConfig config;
  protected final WebDriverWait wait;
  private final SweetAlertComponent sweetAlert;

  public BasePage(WebDriver driver) {
    this.driver = driver;
    this.config = AppConfig.get();
    this.wait = new WebDriverWait(driver, config.expectTimeout());
    this.sweetAlert = this instanceof SweetAlertComponent ? null : new SweetAlertComponent(driver);
  }

  protected void visit(String path) {
    String target = path.startsWith("http") ? path : config.baseUrl() + normalizePath(path);
    driver.navigate().to(target);
    waitForDomReady();
  }

  protected By css(String selector) {
    return By.cssSelector(selector);
  }

  protected By anyCss(String... selectors) {
    return By.cssSelector(String.join(", ", selectors));
  }

  protected By name(String name) {
    return By.name(name);
  }

  protected By inputByName(String name) {
    return By.cssSelector("[name='" + name + "']");
  }

  protected By inputById(String id) {
    return By.cssSelector("#" + id);
  }

  protected By linkByHref(String href) {
    return By.cssSelector("a[href='" + href + "']");
  }

  protected By linkByHrefContains(String hrefPart) {
    return By.cssSelector("a[href*='" + hrefPart + "']");
  }

  protected By buttonByLooseText(String text) {
    return By.xpath(
        "//button[contains(translate(normalize-space(.), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), '"
            + normalizeLooseText(text)
            + "')]");
  }

  protected By actionButton(String action) {
    return actionButton(By.cssSelector("body"), action);
  }

  protected By actionButton(By scope, String action) {
    String cssSelector =
        switch (action) {
          case "view" ->
              ".btn-view, .btn-action.btn-view, [data-action='view'], [title*='Xem'], [title*='Chi tiet'], [title*='detail']";
          case "edit" ->
              ".btn-edit, .btn-action.btn-edit, [data-action='edit'], [title*='Chinh sua'], [title*='Sua'], [title*='Edit']";
          case "delete" ->
              ".btn-delete, .btn-action.btn-delete, [data-action='delete'], [title*='Xoa'], [title*='Delete']";
          case "pay" ->
              ".btn-pay, .btn-action.btn-pay, [data-action='pay'], [title*='Thanh toan'], [title*='Payment']";
          case "approve" ->
              ".btn-approve, .btn-action.btn-approve, [data-action='approve'], [title*='Duyet'], [title*='Approve']";
          case "reject" ->
              ".btn-reject, .btn-action.btn-reject, [data-action='reject'], [title*='Tu choi'], [title*='Reject']";
          default -> throw new IllegalArgumentException("Unsupported action: " + action);
        };

    String scopeExpression = scope.toString().replace("By.cssSelector: ", "");
    if (!scope.toString().startsWith("By.cssSelector: ")) {
      return By.cssSelector(cssSelector);
    }
    return By.cssSelector(scopeExpression + " " + cssSelector);
  }

  protected WebElement visible(By locator) {
    return wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
  }

  protected List<WebElement> visibles(By locator) {
    return driver.findElements(locator).stream().filter(WebElement::isDisplayed).toList();
  }

  protected WebElement firstVisible(By locator) {
    waitForCondition(() -> !visibles(locator).isEmpty(), "No visible element found for " + locator);
    return visibles(locator).get(0);
  }

  protected void click(By locator) {
    wait.until(ExpectedConditions.elementToBeClickable(locator)).click();
  }

  protected void click(WebElement element) {
    wait.until(ExpectedConditions.elementToBeClickable(element)).click();
  }

  protected void fill(By locator, String value) {
    WebElement element = wait.until(ExpectedConditions.elementToBeClickable(locator));
    element.clear();
    element.sendKeys(value);
  }

  protected void fill(WebElement element, String value) {
    wait.until(ExpectedConditions.elementToBeClickable(element));
    element.clear();
    element.sendKeys(value);
  }

  protected void selectByValue(By locator, String value) {
    new Select(visible(locator)).selectByValue(value);
  }

  protected void setInputValue(By locator, String value) {
    WebElement element = visible(locator);
    if (element.isEnabled()) {
      element.clear();
      element.sendKeys(value);
      return;
    }
    ((JavascriptExecutor) driver)
        .executeScript(
            "arguments[0].value = arguments[1]; arguments[0].dispatchEvent(new Event('input', { bubbles: true })); arguments[0].dispatchEvent(new Event('change', { bubbles: true }));",
            element,
            value);
  }

  protected String value(By locator) {
    return visible(locator).getAttribute("value");
  }

  protected String text(By locator) {
    return visible(locator).getText();
  }

  protected String looseText(By locator) {
    return normalizeLooseText(text(locator));
  }

  protected int count(By locator) {
    return driver.findElements(locator).size();
  }

  protected boolean isVisible(By locator) {
    try {
      return driver.findElements(locator).stream().anyMatch(WebElement::isDisplayed);
    } catch (RuntimeException error) {
      return false;
    }
  }

  protected void waitForVisible(By locator) {
    wait.until(ExpectedConditions.visibilityOfElementLocated(locator));
  }

  protected void waitForHidden(By locator) {
    wait.until(ExpectedConditions.invisibilityOfElementLocated(locator));
  }

  protected void waitForUrlContains(String path) {
    wait.until(ExpectedConditions.urlContains(path));
  }

  protected void waitForUrlMatches(String regex) {
    wait.until(ExpectedConditions.urlMatches(regex));
  }

  protected void waitForText(By locator, String expected) {
    waitForCondition(() -> driver.findElements(locator).stream().anyMatch(el -> el.getText().contains(expected)),
        "Text not found for " + locator + ": " + expected);
  }

  protected void waitForTextMatches(By locator, String regex) {
    waitForCondition(() -> driver.findElements(locator).stream().anyMatch(el -> el.getText().matches("(?is).*" + regex + ".*")),
        "Text regex not found for " + locator + ": " + regex);
  }

  protected void waitForValue(By locator, String expected) {
    waitForCondition(() -> expected.equals(driver.findElement(locator).getAttribute("value")),
        "Value not found for " + locator + ": " + expected);
  }

  protected void waitForEnabled(By locator) {
    wait.until(ExpectedConditions.elementToBeClickable(locator));
  }

  protected void waitForDisabled(By locator) {
    waitForCondition(() -> !driver.findElement(locator).isEnabled(), "Locator did not become disabled: " + locator);
  }

  protected void waitForEmpty(By locator) {
    waitForCondition(() -> text(locator).isBlank(), "Locator did not become empty: " + locator);
  }

  protected void waitForCondition(BooleanSupplier predicate, String message) {
    try {
      new WebDriverWait(driver, config.expectTimeout()).until(ignored -> predicate.getAsBoolean());
    } catch (TimeoutException error) {
      throw new AssertionError(message, error);
    }
  }

  protected void waitForDomReady() {
    new WebDriverWait(driver, config.navigationTimeout())
        .until(
            ignored ->
                "complete".equals(((JavascriptExecutor) driver).executeScript("return document.readyState"))
                    || "interactive".equals(((JavascriptExecutor) driver).executeScript("return document.readyState")));
  }

  protected String normalizeLooseText(String value) {
    if (value == null) {
      return "";
    }
    return TextNormalizeHelper.normalizeLooseText(value);
  }

  protected String currentPath() {
    return URI.create(driver.getCurrentUrl()).getPath();
  }

  protected void waitForCurrentUrlContains(String value) {
    waitForUrlContains(value);
  }

  protected void check(By locator) {
    WebElement element = visible(locator);
    if (!element.isSelected()) {
      element.click();
    }
  }

  protected String normalizePath(String path) {
    return path.startsWith("/") ? path : "/" + path;
  }

  protected String urlEncode(String value) {
    return URLEncoder.encode(value, StandardCharsets.UTF_8);
  }

  protected void dismissSweetAlertIfPresent() {
    By confirm = By.cssSelector(".swal2-confirm");
    if (isVisible(confirm) && sweetAlert != null) {
      sweetAlert.confirmIfPresent();
    }
  }

  protected void waitForSweetAlertContainsText(String regex) {
    if (sweetAlert == null) {
      throw new IllegalStateException("Nested SweetAlert calls are not supported.");
    }
    sweetAlert.waitForTextLoose(regex);
  }
}

File: automation-test-playwright/pages/core/BasePage.ts
import { type Locator, type Page } from "@playwright/test";
import { env } from "@helpers-runtime/env";
import { WaitHelper } from "@helpers-browser/WaitHelper";
import { TextNormalizeHelper } from "@helpers-text/TextNormalizeHelper";
import { SweetAlertComponent } from "../components/SweetAlertComponent";

/**
 * Foundation layer for all Page Objects in the framework.
 *
 * Provides common utilities for locator creation, element waits, SweetAlert
 * interaction, text normalization, and page navigation that every page object
 * can inherit and reuse.
 */
export class BasePage {
  protected readonly page: Page;
  protected readonly sweetAlertComponent: SweetAlertComponent;

  /**
   * Creates a new page object bound to the given Playwright Page instance.
   */
  constructor(page: Page) {
    this.page = page;
    this.sweetAlertComponent = new SweetAlertComponent(page);
  }

  /**
   * Navigates to the requested path.
   */
  async visit(path: string): Promise<void> {
    await this.page.goto(path, { waitUntil: "domcontentloaded", timeout: env.navigationTimeout });
  }

  locator(selector: string): Locator {
    return this.page.locator(selector);
  }

  anyLocator(...selectors: string[]): Locator {
    return this.page.locator(selectors.join(", "));
  }

  visible(locator: Locator): Locator {
    return locator.filter({ visible: true });
  }

  firstVisible(locator: Locator): Locator {
    return this.visible(locator).nth(0);
  }

  lastVisible(locator: Locator): Locator {
    return this.visible(locator).nth(-1);
  }

  testId(id: string): Locator {
    return this.page.getByTestId(id);
  }

  inputByName(name: string): Locator {
    return this.page.locator(`[name="${name}"]`);
  }

  inputById(id: string): Locator {
    return this.page.locator(`#${id}`);
  }

  buttonByText(text: string): Locator {
    return this.page.getByRole("button", { name: new RegExp(this.escapeRegExp(text), "i") });
  }

  actionButton(scope: Locator, action: "view" | "edit" | "delete" | "pay" | "approve" | "reject"): Locator {
    const accessibleNameByAction: Record<typeof action, RegExp> = {
      view: /view|detail|xem|chi tiet/i,
      edit: /edit|update|sua|chinh/i,
      delete: /delete|remove|xoa/i,
      pay: /pay|payment|thanh toan/i,
      approve: /approve|duyet/i,
      reject: /reject|tu choi/i
    };
    const cssSelectorByAction: Record<typeof action, string> = {
      view: ".btn-view, .btn-action.btn-view, [data-action='view'], [title*='Xem'], [title*='Chi tiet'], [title*='detail']",
      edit: ".btn-edit, .btn-action.btn-edit, [data-action='edit'], [title*='Chinh sua'], [title*='Sua'], [title*='Edit']",
      delete: ".btn-delete, .btn-action.btn-delete, [data-action='delete'], [title*='Xoa'], [title*='Delete']",
      pay: ".btn-pay, .btn-action.btn-pay, [data-action='pay'], [title*='Thanh toan'], [title*='Payment']",
      approve: ".btn-approve, .btn-action.btn-approve, [data-action='approve'], [title*='Duyet'], [title*='Approve']",
      reject: ".btn-reject, .btn-action.btn-reject, [data-action='reject'], [title*='Tu choi'], [title*='Reject']"
    };
    return scope
      .getByTestId(new RegExp(`${action}|${action}-button|button-${action}`, "i"))
      .or(scope.locator(cssSelectorByAction[action]))
      .or(scope.getByRole("button", { name: accessibleNameByAction[action] }))
      .or(scope.getByRole("link", { name: accessibleNameByAction[action] }))
      .filter({ visible: true })
      .first();
  }

  linkByText(text: string): Locator {
    return this.page.getByRole("link", { name: new RegExp(this.escapeRegExp(text), "i") });
  }

  linkByHref(href: string): Locator {
    return this.page.locator(`a[href="${href}"]`);
  }

  modalById(id: string): Locator {
    return this.page.locator(`#${id}`);
  }

  toastPopup(): Locator {
    return this.page.locator(".swal2-popup").filter({ visible: true }).first();
  }

  async setInputValue(locator: Locator, value: string): Promise<void> {
    const isVisible = await locator.isVisible().catch(() => false);
    const isEditable = await locator.isEditable().catch(() => false);
    if (isVisible && isEditable) {
      await locator.fill(value);
      return;
    }

    await locator.evaluate((element, nextValue) => {
      const input = element as HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement;
      input.value = String(nextValue);
      input.dispatchEvent(new Event("input", { bubbles: true }));
      input.dispatchEvent(new Event("change", { bubbles: true }));
    }, value);
  }

  protected escapeRegExp(value: string): string {
    return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }

  protected normalizeLooseText(value: string): string {
    return TextNormalizeHelper.normalizeLooseText(value);
  }

  async locatorLooseText(locator: Locator): Promise<string> {
    return this.normalizeLooseText((await locator.textContent()) ?? "");
  }

  async waitForTitleContainsLoose(...expectedParts: string[]): Promise<void> {
    const normalizedExpected = expectedParts.map((part) => this.normalizeLooseText(part));
    await WaitHelper.until(
      async () => {
        const normalizedTitle = this.normalizeLooseText(await this.page.title());
        return normalizedExpected.some((part) => normalizedTitle.includes(part));
      },
      { message: `Title did not contain expected text: ${expectedParts.join(", ")}` }
    );
  }

  async waitForPath(pathPattern: RegExp | string): Promise<void> {
    if (typeof pathPattern === "string") {
      await this.waitForPageUrl(new RegExp(pathPattern.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")));
      return;
    }

    await this.waitForPageUrl(pathPattern);
  }

  async waitForPageUrl(pathPattern: RegExp): Promise<void> {
    if (pathPattern.test(this.page.url())) {
      return;
    }

    await this.page.waitForURL(pathPattern);
  }

  async waitForToastMessage(text: string): Promise<void> {
    await this.waitForLocatorText(this.toastPopup(), text);
  }

  async waitForSweetAlertContainsText(text: string | RegExp): Promise<void> {
    await this.sweetAlertComponent.waitForTextLoose(text);
  }

  async confirmSweetAlert(): Promise<void> {
    await this.sweetAlertComponent.waitForVisible();
    await this.sweetAlertComponent.confirm();
  }

  async cancelSweetAlert(): Promise<void> {
    await this.sweetAlertComponent.waitForVisible();
    await this.sweetAlertComponent.cancel();
  }

  /**
   * Dismisses sweet alert if present when visible.
   */
  async dismissSweetAlertIfPresent(): Promise<void> {
    await this.sweetAlertComponent.confirmIfPresent();
  }

  async waitForVisible(locator: Locator): Promise<void> {
    await locator.waitFor({ state: "visible", timeout: env.expectTimeout });
  }

  async waitForHidden(locator: Locator): Promise<void> {
    await locator.waitFor({ state: "hidden", timeout: env.expectTimeout });
  }

  async waitForLocatorText(locator: Locator, text: string | RegExp): Promise<void> {
    await WaitHelper.until(
      async () => {
        const rawText = (await locator.textContent()) ?? "";
        return typeof text === "string" ? rawText.includes(text) : text.test(rawText);
      },
      { message: `Locator did not contain expected text: ${String(text)}` }
    );
  }

  async waitForLocatorValue(locator: Locator, value: string): Promise<void> {
    await WaitHelper.until(async () => (await locator.inputValue().catch(() => "")) === value, {
      message: `Locator did not have expected value: ${value}`
    });
  }

  async waitForLocatorCount(locator: Locator, count: number): Promise<void> {
    await WaitHelper.until(async () => (await locator.count()) === count, {
      message: `Locator count did not become ${count}`
    });
  }

  async waitForEnabled(locator: Locator): Promise<void> {
    await WaitHelper.until(async () => locator.isEnabled().catch(() => false), {
      message: "Locator did not become enabled."
    });
  }

  async waitForDisabled(locator: Locator): Promise<void> {
    await WaitHelper.until(async () => locator.isDisabled().catch(() => false), {
      message: "Locator did not become disabled."
    });
  }

  async waitForEmpty(locator: Locator): Promise<void> {
    await WaitHelper.until(async () => ((await locator.textContent().catch(() => "")) ?? "").trim() === "", {
      message: "Locator did not become empty."
    });
  }

  async waitForCondition(predicate: () => Promise<boolean> | boolean, message?: string): Promise<void> {
    await WaitHelper.until(predicate, { message });
  }
}
